Extension/src/LanguageServer/references.ts

```diff
@@ -218,58 +235,65 @@ export class ProgressHandler {
+            this.referencesProgressMethod = (progress: vscode.Progress<{message?: string; increment?: number }>, token: vscode.CancellationToken) =>
+            // tslint:disable-next-line: promise-must-complete
+                new Promise((resolve) => {
+                    this.reportProgress(progress, true);
+                    let currentUpdateProgressTimer: NodeJS.Timeout = setInterval(() => {
+                        if (token.isCancellationRequested && !this.referencesCanceled) {
+                            this.client.cancelReferences();
+                            this.client.sendRequestReferences();
+                            this.referencesCanceled = true;
+                        }
+                        if (this.referencesCurrentProgress.referencesProgress === ReferencesProgress.Finished || this.referencesCurrentProgressUICounter !== referencePreviousProgressUICounter) {
+                            clearInterval(currentUpdateProgressTimer);
+                            if (this.referencesCurrentProgressUICounter !== referencePreviousProgressUICounter) {
+                                referencePreviousProgressUICounter = this.referencesCurrentProgressUICounter;
+                                this.referencesPrevProgressIncrement = 0; // Causes update bar to not reset.
+                                vscode.window.withProgress(this.referencesProgressOptions, this.referencesProgressMethod);
+                            }
+                            resolve();
+                        } else {
+                            this.reportProgress(progress, false);
+                        }
+                    }, this.referencesProgressUpdateInterval);
+                });
+            vscode.window.withProgress(this.referencesProgressOptions, this.referencesProgressMethod);
+            clearInterval(this.referencesDelayProgress);
+        }, this.referencesProgressDelayInterval);
+    }
+
     public handleProgress(notificationBody: ReportReferencesProgressNotification): void {
+        this.initializeViews();
+
         switch (notificationBody.referencesProgress) {
             case ReferencesProgress.StartedRename:
             case ReferencesProgress.Started:
-                this.referencesStartedWhileTagParsing = this.client.IsTagParsing;
-                let mode: ReferencesCommandMode = notificationBody.referencesProgress === ReferencesProgress.StartedRename ? ReferencesCommandMode.Rename :
-                    (this.visibleRangesDecreased && (Date.now() - this.visibleRangesDecreasedTicks < this.ticksForDetectingPeek) ?
-                    ReferencesCommandMode.Peek : ReferencesCommandMode.Find);
-                this.client.setReferencesCommandMode(mode);
                 if (this.client.ReferencesCommandMode === ReferencesCommandMode.Peek) {

```

Comment:
Why is this not in the handleProgressStarted method?

Reply:
I wanted to separate telemetry stuff from the logic that processes the data.